[SPARK-23166][ML] Add maxDF Parameter to CountVectorizer

ymazari opened the pull request:
## What changes were proposed in this pull request?
Currently, the CountVectorizer has a minDF parameter.

It might be useful to also have a maxDF parameter.
It will be used as a threshold for filtering all the terms that occur very frequently in a text corpus, because they are not very informative or could even be stop-words.

This is analogous to scikit-learn, CountVectorizer, max_df.

Other changes:
- Refactored code to invoke "filter()" conditioned on maxDF or minDF set.
- Refactored code to unpersist input after counting is done.

## How was this patch tested?
Unit tests.

ymazari committed:
[SPARK-23166][ML] Add maxDF Parameter to CountVectorizer

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

ymazari commented:
@hhbyyh, @MLnick, @jkbradley could you please help reviewing this? Thanks.

ymazari commented:
@mgaido91 I closed the previous PR and opened this one.
I am answering your comments here.

ymazari commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -169,7 +201,7 @@ class CountVectorizer @Since("1.5.0") (@Since("1.5.0") override val uid: String)
       (wc1 + wc2, df1 + df2)
     }.filter { case (word, (wc, df)) =>
-      df >= minDf
+      (df >= minDf) && (df <= maxDf)
```
> from @mgaido91: nit: the parenthesis are not needed
Right. I added them for the purpose of clarity.

ymazari commented on mllib/src/test/scala/org/apache/spark/ml/feature/CountVectorizerSuite.scala:
```diff
@@ -119,6 +119,41 @@ class CountVectorizerSuite extends SparkFunSuite with MLlibTestSparkContext
+
+    cvModel2.transform(df).select("features", "expected").collect().foreach {
+      case Row(features: Vector, expected: Vector) =>
+        assert(features ~== expected absTol 1e-14)
```
> by @mgaido91: may you please also add a UT to check that setting both maxDF and minDF works as expected?

Yes, I will.

ymazari replied:
Done.

ymazari commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -113,7 +132,11 @@ private[feature] trait CountVectorizerParams extends Params with HasInputCol wit
-  setDefault(vocabSize -> (1 << 18), minDF -> 1.0, minTF -> 1.0, binary -> false)
+  setDefault(vocabSize -> (1 << 18),
+    minDF -> 1.0,
+    maxDF -> Long.MaxValue,
```
> by @mgaido91: what about avoiding to set a default value and apply the filter only if it was set?

That'd be easy to do, but would make the code a couple lines longer.
I will do it if you think it's really better.

mgaido91 replied:
Let's hear also what other people think about this. Probably in practice this is ok, but the other solution is cleaner IMHO.

srowen replied:
Yeah, I get it. It didn't work that way before; seems valuable only if you can avoid the whole filter stage (both values are set to filter nothing). Even there, I wonder if that makes any appreciable difference? I suppose I would have just done the straightforward thing here, and so this change looks OK to me.

ymazari committed:
[SPARK-23166][ML] Add maxDF Parameter to CountVectorizer
- More unit tests for when both minDF and maxDF are used.

ymazari commented:
@srowen It seems that this PR needs Admin approval. Could you please help getting it to the next step? Thanks.

SparkQA commented:
**[Test build #4077 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4077/testReport)** for PR 20367 at commit [`daa0870`](https://github.com/apache/spark/commit/daa08705b2cb8e097c037171b809f0cbfab3f52c).

SparkQA commented:
**[Test build #4077 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4077/testReport)** for PR 20367 at commit [`daa0870`](https://github.com/apache/spark/commit/daa08705b2cb8e097c037171b809f0cbfab3f52c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -160,6 +187,11 @@ class CountVectorizer @Since("1.5.0") (@Since("1.5.0") override val uid: String)
+      $(maxDF)
+    } else {
+      $(maxDF) * input.cache().count()
+    }
```
I wonder if it's best to check that the max is >= the min? this may be the only place that can be done as one might be a fraction and the other not.

Also this ends up counting the input twice when both params are set, which might be worth avoiding. It could be computed and saved only when needed.

(I suppose this also fails to unpersist `input`, which could happen right after `wordCounts.count()`. That's a slightly different issue, so not too concerned about whether to fix it here or not.)

ymazari replied:
Good points.
- I added a check that maxDF >= minDF
- Changed the code so that counting (and caching) is done only once
- I refactored the code so that "filter()" is only invoked if minDF or maxDF is set
- Added an un-persisting the input after the counting is done.

ymazari committed:
[SPARK-23166][ML] Add maxDF Parameter to CountVectorizer
- Added check that maxDF >= minDF
- Performed input caching and count only once
- Refactored code to invoke "filter()" conditioned on maxDF or minDF set.
- Refactored code to unpersist input after counting is done.

mgaido91 commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -155,24 +182,47 @@ class CountVectorizer @Since("1.5.0") (@Since("1.5.0") override val uid: String)
     transformSchema(dataset.schema, logging = true)
     val vocSize = $(vocabSize)
     val input = dataset.select($(inputCol)).rdd.map(_.getAs[Seq[String]](0))
+    val filteringRequired = isSet(minDF) || isSet(maxDF)
```
this would cause a count also when it is not necessary, ie. when `minDF`/`maxDF` are set not as a fraction but as an integer

srowen replied:
I imagine you just want the same conditions that are used below, like `$(minDF) >= 1.0`

mgaido91 replied:
yes, this is the right condition to choose whether  to `filter` or not, but the right one to choose whether to do the count or not is the one you just stated

ymazari replied:
Changed.

mgaido91 commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -155,24 +182,47 @@ class CountVectorizer @Since("1.5.0") (@Since("1.5.0") override val uid: String)
       $(minDF)
     } else {
-      $(minDF) * input.cache().count()
+      $(minDF) * maybeInputSize.getOrElse(1L)
```
I think `get` is enough...when we are here it must be set

ymazari replied:
Right. I guess I was overdoing it here.

mgaido91 commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -155,24 +182,47 @@ class CountVectorizer @Since("1.5.0") (@Since("1.5.0") override val uid: String)
+    val maxDf = if ($(maxDF) >= 1.0) {
+      $(maxDF)
+    } else {
+      $(maxDF) * maybeInputSize.getOrElse(1L)
```
ditto

ymazari replied:
Changed.

mgaido91 commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -155,24 +182,47 @@ class CountVectorizer @Since("1.5.0") (@Since("1.5.0") override val uid: String)
+    }
+
+    val wordCounts = maybeFilteredWordCounts
+      .map { case (word, (count, dfCount)) => (word, count) }
```
nit: `case (word, (count, _))`

ymazari replied:
Changed.

mgaido91 commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -155,24 +182,47 @@ class CountVectorizer @Since("1.5.0") (@Since("1.5.0") override val uid: String)
+    }
+
+    val maybeFilteredWordCounts = if (filteringRequired) {
+      allWordCounts.filter { case (word, (wc, df)) => (df >= minDf) && (df <= maxDf) }
```
nit: `case (_, (_, dfCount))` and I still would prefer to avoid the unnecessary parenthesis

ymazari replied:
Changed.

mgaido91 commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -155,24 +182,47 @@ class CountVectorizer @Since("1.5.0") (@Since("1.5.0") override val uid: String)
-    }.cache()
+    }
+
+    val maybeFilteredWordCounts = if (filteringRequired) {
```
after addressing my previous comment this would need to be changed

ymazari replied:
Done.

srowen reviewed (changes requested):
Looking good, better than the original code

ymazari committed:
[SPARK-23166][ML] Add maxDF Parameter to CountVectorizer

ymazari commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -155,24 +182,48 @@ class CountVectorizer @Since("1.5.0") (@Since("1.5.0") override val uid: String)
-    }.cache()
+    }
+
+    val filteringRequired = isSet(minDF) || isSet(maxDF)
```
Making a variable here for the sake of clarity.

mgaido91 commented:
LGTM, thanks

SparkQA commented:
**[Test build #4083 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4083/testReport)** for PR 20367 at commit [`c62d9a5`](https://github.com/apache/spark/commit/c62d9a51ae885204e76fc6997952aaddb5aeb64e).

SparkQA commented:
**[Test build #4083 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4083/testReport)** for PR 20367 at commit [`c62d9a5`](https://github.com/apache/spark/commit/c62d9a51ae885204e76fc6997952aaddb5aeb64e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
Thanks @ymazari this is a nice improvement! merged to master
